DOCS-6409-http-THROWS-descriptions-kt (#1577)
mulesoft/docs-connectors: http/1.5/modules/ROOT/pages/http-troubleshooting.adoc

= Troubleshoot HTTP Connector - Mule 4
:keywords: anypoint studio, esb, connectors, http, https, http headers, troubleshooting, rest, raml
:page-aliases: connectors::http/http-troubleshooting.adoc

> To troubleshoot Anypoint Connector for HTTP (HTTP Connector), become familiar with the information about troubleshooting connectivity, permissions and server-side issues, troubleshooting Mule application issues, and interpreting commonly thrown messages.

Reviewer: This sentence begins with the goal of troubleshooting, and then lists four types of information that you need to do so: two include "troubleshooting" and two do not. 
Is there a way to make the information list items parallel?
Does this mean the same thing:
```suggestion
To troubleshoot Anypoint Connector for HTTP (HTTP Connector), become familiar with the information about connectivity, permissions and server-side issues, Mule application issues, and commonly thrown messages.
```

Reviewer: Actually, I wonder if this introduction should more overtly introduce the actual user subtasks indicated in the subtitles contained within?
```suggestion
To troubleshoot Anypoint Connector for HTTP (HTTP Connector), enable wire logging and TLS debugging, ensure proper use of the `Host` header and HTTP encoding, and become familiar with commonly thrown messages.
```

== Enable Wire Logging

> Sometimes you are not sure why the application is failing and you require a better understanding of what is going on with the app. Enabling wire logging logs the actual raw HTTP requests and responses flowing. You can the use the logs to validate the current configuration or to compare the HTTP Connector behavior to other clients and servers.

Reviewer: Just a suggestion for minimalism:
```suggestion
Troubleshoot application performance by enabling wire logging, which documents the actual raw HTTP requests and responses in the app workflow. Then use the logs to validate the current configuration or to compare HTTP Connector behavior to that of other clients and servers.
```
Will your audience understand what "behaviors" are referred to here?

To enable HTTP wire logging:

. Access Anypoint Studio and navigate to the *Package Explorer* view.
. Open your application's project name.
. Open the `src/main/resources` path folder.

> . Open the `log4j2.xml` file inside the folder.

Reviewer: Is there a reason to use "open" instead of "click"?
Also, you do not "open" a project name: you click the app name to open the app.

</Configuration>
----

[start=6]
. Save your changes.
. Click the project name in *Package Explorer* and then click *Run* > *Run As* > *Mule Application*.

== Enable TLS Debugging

> When SSL failures occur because of HTTPS, usually connections cannot be established and wire logging is not helpful. In these situations, enable TLS debugging that logs all SSL related data, such as connection attempts, handshakes and negotiations.

Reviewer: Will the audience already know that the issue is an SSL failure?
Do we need to mention HTTPS along with SSL, or relate this task to wire logging?
Does this mean the same thing:
```suggestion
Troubleshoot SSL failures by enabling TLS debugging, which logs SSL-related data such as connection attempts, handshakes, and negotiations.
```

> You can then examine the logs to determine whether the correct certificates are in place, why a cipher or protocol mismatch occurs or why connectivity fails.

Reviewer: If you adopt suggestion at Line 9, reword this (and following) to use the pattern "Troubleshoot <this kind of error> by <doing this>. Then use <the results to fix>.
```suggestion
Then use the logs to determine whether the correct certificates are in place, why a cipher or protocol mismatch occurs, or why connectivity fails.
```
Isn't "why connectivity fails" the main issue, and certificates, mismatch the causes? Do we really want:
```suggestion
Then use the logs to determine the cause of connectivity failure, such as incorrect certificates or cipher or protocol mismatch.
```

To enable TLS debugging, configure the following system property: `-Djavax.net.debug=ssl`.

> == Troubleshoot Host Header Misuse

Reviewer: ```suggestion
== Correct Host Header Misuse
```

HTTP clients use the `Host` header to indicate the host and port number of the resource requested. The HTTP Connector *Request* operation autogenerates this mandatory header considering the configured host and port.

Often, servers do not account for the port information because it's optional, and fail when receiving a request that has both the host and port data.

> Such failures should be reported to the server owners. As a workaround for the issue, modify the *Request* operation by providing an explicit `Host` header without the port section.

Reviewer: Alternative for minimalism?

The HTTP Connector *Request* operation automatically generates a `Host` header that indicates both the host and the port number of a resource requested from a server. However, servers often consider the port information optional and fail if they receive a request containing it.

Report such failures to the server owner to correct. Meanwhile, work around the issue by manually modifying the *Request* operation to output a `Host` header without the port information.

> The following examples shows the `<http:headers>` configuration that sends a `Host: somehost.com` header:

Reviewer: ```suggestion
The following example shows a `<http:headers>` configuration that sends a `Host: somehost.com` header:
```

Without the explicit header configuration, an autogenerated `Host: somehost.com:8081` would be sent.

> == Troubleshoot HTTP Encoding

Reviewer: ```suggestion
== Correct HTTP Encoding
```

Reviewer: I suggest similar rewording here, to avoid "there is" (per style ref) and to minimize:

To obtain better performance when working with streamed data, HTTP Connector sends...encoding when the size of the data being sent is unknown. However, some HTTP clients and servers...fail.

Report such failures to the server or client owner to correct. Meanwhile, work around the issue by manually modifying the *Listener* and *Request* operations to output `Content-Length` encoded bodies.

</flow>
----

> The following example shows the HTTP *Requester* operation configured with `requestStreamingMode` `NEVER`, which indicates to use `Content-Length` encoding:

Reviewer: ```suggestion
The following example shows an HTTP *Requester* operation configured with `requestStreamingMode` `NEVER` to use `Content-Length` encoding:
```

</flow>
----

== Understand Common Throws

> Here is a list of common throws messages and how to interpret them.

Reviewer: ```suggestion
Here is a list of common throws messages and how to interpret them:
```

* HTTP:BAD_REQUEST

>  The server can not understand the request due to invalid syntax. For example, when some header size exceeds the maximum.

Reviewer: ```suggestion
 The server cannot understand the request due to invalid syntax, for example, when a header size exceeds the maximum.
```

* HTTP:CLIENT_SECURITY

>  Indicates a security type problem enforced by an external entity.

Reviewer: Reword to make this description parallel with the others (noun-based rather than verb-based):
```suggestion
 A security problem enforced by an external entity occurred.
```

* HTTP:CONNECTIVITY

 A problem occurred and a connection could not be established.

* HTTP:FORBIDDEN

>   The server refuses to give the requested resource because the client does not have access rights to the content.

Reviewer: ```suggestion
  The server refuses to send the requested resource because the client does not have access rights to the content.
```

* HTTP:INTERNAL_SERVER_ERROR

>  The server encountered a situation which prevented from fulfilling the request.

Reviewer: ```suggestion
 The server encountered a situation that prevented it from fulfilling the request.
```

* HTTP:METHOD_NOT_ALLOWED

>  The request method is known by the server but the target resource does not support it.

Reviewer: ```suggestion
 The server request method is not supported by the target resource.
```

* HTTP:NOT_ACCEPTABLE

 The server cannot produce a response matching the list of acceptable values defined in the request's proactive content negotiation headers.

* HTTP:NOT_FOUND

 The server could not find the requested resource.

* HTTP:PARSING

>  [DEPRECATED] Kept for compatibility. The thrown message returned in a removed parsing mechanism.

Reviewer: ```suggestion
 [DEPRECATED but kept for compatibility.] The parsing mechanism has been removed.
```

* HTTP:RETRY_EXHAUSTED

 The maximum number of retries for the operation is reached.

* HTTP:SECURITY

 The requester authentication failed.

* HTTP:SERVICE_UNAVAILABLE

 The server is unable to manage the request because it is down for maintenance or overloaded.

* HTTP:TIMEOUT

 The request sent by an http:requester timed out.

* HTTP:TOO_MANY_REQUESTS

 Too many request were sent in a given amount of time.

* HTTP:UNAUTHORIZED

>  Authentication failed or has not yet been provided to get the requested response.

Reviewer: ```suggestion
 Request authentication failed or has not yet been provided.
```

* HTTP:UNSUPPORTED_MEDIA_TYPE

 The server does not support the media format of the requested data.

* HTTP:BAD_GATEWAY

>  The server acting as a gateway or proxy to manage the request got an invalid response.

Reviewer: ```suggestion
 The server acting as a gateway or proxy to manage the request received an invalid response.
```

* HTTP:GATEWAY_TIMEOUT

>  The server acting as a gateway or proxy to manage the request did not receive a response in time.

Reviewer: ```suggestion
 The server acting as a gateway or proxy to manage the request did not receive a response within a specified time.
```

* HTTP:BASIC_AUTHENTICATION

>  HTTP Request operation requires basic authentication to send requests to the service, or the provided credentials are incorrect.

Reviewer: ```suggestion
 Either the HTTP *Request* operation lacks basic authentication to send requests to the service, or the provided credentials are incorrect.
```
